scripts/main/contextMenu.js

```diff
@@ -474,23 +524,74 @@ contextMenu.photoMore = function (photoID, e) {
 	basicContext.show(items, e.originalEvent);
 };
 
+/**
+ * @param {Album[]} albums
+ * @param {string} albumID
+ *
+ * @returns {string[]}
+ */
 contextMenu.getSubIDs = function (albums, albumID) {
-	let ids = [parseInt(albumID, 10)];
-	let a;
+	let ids = [albumID];
 
-	for (a = 0; a < albums.length; a++) {
-		if (parseInt(albums[a].parent_id, 10) === parseInt(albumID, 10)) {
-			ids = ids.concat(contextMenu.getSubIDs(albums, albums[a].id));
+	albums.forEach(function (album) {
+		if (album.parent_id === albumID) {
+			ids = ids.concat(contextMenu.getSubIDs(albums, album.id));
 		}
 
-		if (albums[a].albums && albums[a].albums.length > 0) {
-			ids = ids.concat(contextMenu.getSubIDs(albums[a].albums, albumID));
+		if (album.albums && album.albums.length > 0) {
+			ids = ids.concat(contextMenu.getSubIDs(album.albums, albumID));
 		}
-	}
+	});
 
 	return ids;
 };
 
+/**
+ * @callback TargetAlbumSelectedCB
+ *
+ * Called by {@link contextMenu.move} after the user has selected a target ID.
+ * In most cases, {@link album.setAlbum} or {@link photo.setAlbum} are
+ * directly used as the callback.
+ * This design decision is the only reason, why this callback gets more
+ * parameters than the selected target ID.
+ * The parameter signature of this callback matches {@link album.setAlbum}.
+ *
+ * However, the callback should actually enclose all other parameters it
+ * needs and only receive the target ID.
+ *
+ * TODO: Re-factor callbacks.
+ *
+ * @param {string[]} IDs      the source IDs
+ * @param {?string} targetID  the ID of the target album
+ * @param {boolean} [confirm] indicates whether the callback shall show a
+ *                            confirmation dialog to the user for whatever to
+ *                            callback is going to do
+ * @returns {void}
+ */
+
+/**
+ * Shows the context menu with the album tree and allows the user to select a target album.
+ *
+ * **ATTENTION:** The name `move` of this method is very badly chosen.
+ * The method does not move anything, but only shows the menu and reports
+ * the selected album.
+ * In particular, the method is used by any operation which needs an target
+ * album (i.e. merge, copy-to, etc.)
+ *
+ * TODO: Find a better name for this function.
+ *
+ * The method calls the provided callback after the user has selected a
+ * target album and passes the ID of the target album together with the
+ * source `IDs` and the event `e` to the callback.
+ *
+ * TODO: Actually the callbacks should enclose all additional parameters (e.g., `IDs`) they need. Refactor the callbacks.
+ *
+ * @param {string[]} IDs - IDs of source objects (either album or photo IDs)
+ * @param {jQuery.Event} e - Some (?) event
+ * @param {TargetAlbumSelectedCB} callback - to be called after the user has selected a target ID
+ * @param {string} [kind=UNSORTED] - Name of root album; either "UNSORTED" or "ROOT"; TODO: Why do we need two different names for the same thing?
```

Comment:
spaghetti is the answer. :/

Reply:
Well, yeah, it's ugly, but there actually _is_ a reason :smile:. `UNSORTED` is used when moving photos, `ROOT` when moving albums. They result in different strings in the UI: `Unsorted` and `Albums`, respectively.

Reply:
> Well, yeah, it's ugly, but there actually _is_ a reason :smile:. `UNSORTED` is used when moving photos, `ROOT` when moving albums. They result in different strings in the UI: `Unsorted` and `Albums`, respectively.

Maybe I am too dumb, but I don't get. Well, of course I understand that the different constants result into different string on the UI. Fine. But _why_ do we need it? What is the purpose of calling the root album "unsorted" when moving photos and "albums" when copying photos? Isn't that confusing?

Reply:
Because that's what the front end UI presents the users with. Photos without an album are not displayed at the top-level; they are instead displayed in a separate Unsorted smart album. So that's one of the targets we present to the users when moving/copying photos. It's different for albums since they _are_ displayed at the top level, so for them we display "Albums" (there was a time when we used to display "Root" I believe, but I thought it was confusing to end-users so I renamed it).

You may remember that **I** was the one who recently advocated for cleaning up this mess by getting rid of the whole Unsorted smart album and either supporting the displaying of photos at the top level, or requiring that all photos are in albums. But you guys didn't agree with me.

Reply:
> You may remember that I was the one who recently advocated for cleaning up this mess

Yes, I do remember. :-)

And thank you again, _this_ is the kind explanation I was missing. I will make your answer part of the jsDoc comment of the that method, because your answer contains the kind of information which really helps a programmer to understand the code. So the point is not that the root album is named differently (that is only a circumstantial consequence), the actual point is that in one case the root album (aka "unsorted") can be picked as a valid target (for photos) while in other cases the root album cannot be picked as a target (for albums).

I won't change the code, because this is not the concern of my upcoming re-factoring, but IHMO the parameter is named completely confusing and misleading. Why it is called `kind`, the unaware programmer wonders kind of what? IHMO the parameter should be a boolean and be called `isRootSelectableTarget`. That would really convey its primary meaning. The different names is just a side-effect.

Reply:
> So the point is not that the root album is named differently (that is only a circumstantial consequence), the actual point is that in one case the root album (aka "unsorted") can be picked as a valid target (for photos) while in other cases the root album cannot be picked as a target (for albums).

Erm, no? We _can_ move a subalbum to the root level. This function actually _does_ take a separate `display_root` argument though I don't remember off the top of my head when we use it. I think `kind` is just about what name to display.

One thing is for sure: this function is overloaded to an unreasonable degree. It's an ugly interface.

Reply:
I fixed the comment.